salt/_modules/metalk8s_checks.py

```
@@ -169,6 +175,194 @@ def services(conflicting_services=None, raises=True, **kwargs):
+            roles = (
+                __pillar__.get("metalk8s", {})
+                .get("nodes", {})
+                .get(__grains__["id"], {})
+                .get("roles")
+            )
+
+        # If role not yet set consider we have all roles
+        if roles is None:
+            roles = listening_process_per_role.keys()
+
+        # Compute full dict of listening process according to local `roles`
+        # Note: We consider all minions as "node"
+        listening_process = listening_process_per_role.get("node") or {}
+        for role in roles:
+            listening_process.update(listening_process_per_role.get(role) or {})
+
+    if not isinstance(listening_process, dict):
+        raise ValueError(
+            "Invalid listening process, expected dict but got {}.".format(
+                type(listening_process).__name__
+            )
+        )
+
+    errors = []
+
+    all_listen_connections = __salt__["metalk8s_network.get_listening_processes"]()
+
+    for address, matching in listening_process.items():
+        if isinstance(address, int):
+            address = str(address)
+        ip = None
+        port = address
+        if len(address.rsplit(":")) == 2:
+            ip, port = address.rsplit(":")
+
+        if ip and ip in ["control_plane_ip", "workload_plane_ip"]:
+            ip = __grains__["metalk8s"][ip]
+
+            # We also update the `address` for error message
+            address = "{}:{}".format(ip, port)
+
+        processes = matching.get("expected")
+        if not isinstance(processes, list):
+            processes = [processes]
+
+        if not matching.get("mandatory") and None not in processes:
+            processes.append(None)
+
+        process_on_port = all_listen_connections.get(
+            str(port), all_listen_connections.get(int(port))
+        )
+
+        success = False
+        error_process = {}
+        # If one of this process is expected one, succeed
+        for process in processes:
```

Took me some time to understand this loop :confused:, maybe a comment could help others.

If I understand correctly, this loop tries to find the first successful match, among the `processes` list.

Say we have:

```python
listening_process = {
  "127.0.0.1:6443": {
    "expected": "kube-apiserver",
    "mandatory": False
  }
}
```

In this case, we will have:

```python
processes = ["kube-apiserver", None]
```

So if `kube-apiserver` is listening, we will not hit the `None` option, and everything will go fine.
In essence, setting `"mandatory": False` means having a fallback to "nothing listening" after having tried all `"expected"` patterns.

---

Yes exactly but note also that if you would have put the None manually and got a 
```python
processes = [None, "kube-apiserver"]
```
The result would have been the same, just error for None if apiserver listening but this error is ignored in the resulting message anyway since one of the process match.

I will add a comment